### DISCUSSION TITLE
Wait for OpenVPN to connect before opening RDP Window

### DISCUSSION THREAD

**Discussion by MWinhard** [1Remote/1Remote#363] [Q&A]
One of my windows servers is behind VPN. So I have to connect the VPN tunnel first, wait until it's connected, then run the RDP connection. I'm trying to automate this with PRemoteM 7.2.3 using OpenVPN 2.5.8.

Script before connect: "C:\Program Files\OpenVPN\bin\openvpn-gui.exe"  --command myconnection.ovpn

Script after disconnected: "C:\Program Files\OpenVPN\bin\openvpn-gui.exe" --command disconnect myconnection.ovpn

Both scripts work. However they return immediately. So PRemoteM launches the RDP connection immediately. The result is that the RDP connection doesn't connect.

Any ideas how to fix this?

Markus

**Comment by majkinetor** [comment]
Hm... what happened with replies here... I got them on email...

**Comment by majkinetor** [comment]
@VShawn recommended:


i suggest you can wait in the script file, e.g. in powershell:



```powershell
"C:\Program Files\OpenVPN\bin\openvpn-gui.exe" --command myconnection.ovpn # open vpn
do {
  Write-Host "waiting..."
  sleep 3      
} until(Test-NetConnection $Env:PRM_HOST -Port $Env:PRM_PORT | ? { $_.TcpTestSucceeded } ) # loop until IP:port is responding
```

I did not test the script, you may need debug above scrip to make it works :)

**Reply by majkinetor** [reply]
Something like this is the way to go indeed, u would need to have timeout though, so it doesn't happen to wait forever in the background.

**Comment by MWinhard** [comment]
Great idea! I had to tweak your suggestion a little bit, but now it works. 😁😁😁

```powershell
"C:\Program Files\OpenVPN\bin\openvpn-gui.exe" --command myconnection.ovpn # open vpn tunnel
$loop = 0
do {
  $loop++
  Write-Host "Checking VPN connection... ($loop)"
} until (Test-Port $Env:PRM_HOST $Env:PRM_PORT | ? { $_.PortOpened } ) # loop until IP:port is responding
```

Function Test-Port is from [A fast way to check TCP port in Powershell](https://www.copdips.com/2019/09/fast-tcp-port-check-in-powershell.html)

I had to use Test-Port instead of Test-NetConnection, because
1. Test-NetConnection is kind of caching the result or the ip route or whatever, when running in a loop. So when it was False at first time, it never will become True.
2. Test-NetConnection is incredibly slow while the ip address is not available (because OpenVPN is still connecting). Test-Port is really fast.

**Reply by majkinetor** [reply]
This looks happier:

```ps1
"$Env:ProgramFiles\OpenVPN\bin\openvpn-gui.exe" --command myconnection.ovpn

Write-Host "Checking VPN connection" -NoNewLine
$i = 0
while(1) {
  Write-Host -NoNewLine ' .'
  $portOpened = Test-Port $Env:PRM_HOST $Env:PRM_PORT | ? { $_.PortOpened } 
  if ($portOpened -or $i -gt 60) { break } else { Start-Sleep 1; $i++}
} 
if ($i -gt 60) { Write-Error 'timeout' }
```

**Reply by MWinhard** [reply]
I agree that this is more readable. I'm using your ideas now in my script. And I removed the sleep command as Test-Port itself takes nearly one second while the connection is not yet available.

Unfortunately PRemoteM ignores when a timeout happens. The RDP session is started nevertheless. I also tried Exit 1 instead of Write-Error 'timeout'. Same result.

Does PRemoteM check if my "script before connect" was successful?

**Reply by VShawn** [reply]
Unfortunately nope, PRemoteM will continue opening the session whatever script's exit code is.

I didn't expect a scrip will give a return code when I design this feature.

**Reply by MWinhard** [reply]
It's easy to return an exit code from ps1/cmd/bat files. Use "Exit 1" in powershell or "exit /b 1" in batch files. 0 means success, everything else means failure.

Having said this, it would make sense to implement checking the return value of the "before script" in PRemoteM at some rainy day in the future. Till then I can live with the current behavior. 😉

**Reply by VShawn** [reply]
yeap, it's available on latest version. https://github.com/1Remote/1Remote/issues/376

**Comment by VShawn** [comment]
I have a question.

How do you solve the problem of VPN disconnection? For example, if you have two sessions that require a VPN connection and closing one will cause the VPN to disconnect, thereby causing the other connection to also be disconnected.

**Reply by majkinetor** [reply]
Script can increment on disk number of open connections and disconnect VPN only if there is no more running.

Not sure what will happen if it is forcefully disconnected (i.e. internet problem), does the end script runs ?

**Reply by VShawn** [reply]
Yes I run the script when the connection instance destruct, no matter if it is forcefully disconnected

**Reply by majkinetor** [reply]
So that is solution then, its trivial.

**Reply by majkinetor** [reply]
Other one would be to allow some IPC mechanism to ask running PRM for info about specific connections, but I recon that is more work for both you to develop and user of the script. It would also be probably OS specific while I would suggest to keep stuff cross platform as much as possible.

**Comment by MWinhard** [comment]
Fortunately I don't have to solve this problem. 😊
For all my VPN connections there's only one server behind each VPN connection.
